feat(aws-codebuild): add `enableBatchBuilds()` to Project

tjenkinson opened the pull request:
In order for a CodeBuild to run in batch mode, a batch service role is needed, as described [here in the docs](https://docs.aws.amazon.com/codebuild/latest/userguide/batch-build.html).

>Batch builds introduce a new security role in the batch configuration. This new role is required as CodeBuild must be able to call the StartBuild, StopBuild, and RetryBuild actions on your behalf to run builds as part of a batch. Customers should use a new role, and not the same role they use in their build...

At first I thought lets add this by default, but then I realised when `BatchConfiguration` is set to something, in the aws console the default 'start build' button behaviour changes to start a batch build by default instead :/

So now this adds a new `supportBatchBuildType` option, which when `true` adds minimum amount of `BatchConfiguration` needed for batch builds to run.

I also updated the doc blocks for the webhook option and CodePipeline action option, because users of those also need to set this option. It would be nice to auto-enable this if a webhook or CodeBuild action is configured, but that sounds pretty complicated.

I'm not sure why anyone would need to customise this role, given it appears to only be used internally to do those 3 things, so this PR does not make it configurable. My thinking is that this could be added later if needed, but this PR just gets batch builds working.

In the future if people want control of the other `BatchConfiguration` options I was thinking these could be added and would require `supportBatchBuildType` to be `true`.

related: https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/621

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 20f197ec0e8b2e11bcf24df78e0bacf7b8c5ad83
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bccebbd4-8998-4845-be5d-cac9c2a93282%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8eade4ca09758961734310167617a36f93a9d553
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e612263b-cd7d-4dcf-802e-825e5e845873%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 9981c8a48cf6065f73fcbeed7a71875e359db171
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d636c633-38ba-476f-8d78-6384d01b736e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 88339adbeca42f6deef791d1bca71b0ede1fd734
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8ec38c56-f3e2-46be-a960-44ec09cbb4e7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: e98e12d0fec9072694bef40d798d35c1376d76b2
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=33716a0d-82c1-453b-84d8-a6e0c1ea7a92%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 0ec578464f941ad3480e2c2021faf01311cbb8e2
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9ef3ce24-bc81-4435-ae64-b58c98d2c86b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 reviewed (changes requested):
Thanks a lot for the contribution @tjenkinson !

May I suggest something? 🙂.

Your change currently says, "if you use `webhookTriggersBatchBuild`, you have to use `supportBatchBuildType` too". I would posit that that's actually backwards, and what you want is actually: "if you use `webhookTriggersBatchBuild`, CDK will make sure that that `supportBatchBuildType` is set!" 🙂.

Given that, I'd suggest the following changes:

1. Make the `buildBatchConfig` in the `CfnProject` be set with a Lazy (like `encryptionKey` is, for example).
2. Remove the `supportBatchBuildType` property from `ProjectProps`.
2. Add a method `enableBatchBuilds(): BatchBuildConfig | undefined` to `IProject`. It will return `undefined` for imported Projects, where this cannot be changed. For new Projects, it will create a new Role like you do today, and set the `buildBatchConfig` private property of the L1 so that the above Lazy can set it. `BatchBuildConfig` is a very simple new interface that only contains a `role` property. You should return the newly created Role there, so that customers can update it if necessary. The new method also gives us flexibility in adding an optional `role` argument to it in the future.
3. In the `ThirdPartyGitSource`, if `webhookTriggersBatchBuild` was set to `true`, call the `enableBatchBuilds()` method on the `IProject` instance it gets in `bind()`.
4. Similarly in `CodeBuildAction`: if `executeBatchBuild` was set to `true`, call the `enableBatchBuilds()` method on the `IProject` it gets in the properties.

Let me know if this description is clear, or if you need any help implementing it!

tjenkinson commented:
Makes sense to me. Will update.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 48617a27143fa922c82d8b24636049127499ad10
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4654a62d-ae81-4106-8027-862ab1382bf7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

tjenkinson commented:
@skinny85 updated.

I left the `supportBatchBuildType` option in for now because I think it's something you also might want to enable for triggering a build manually? Otherwise you'd need to construct the project and then immediately call `enableBatchBuilds()`.

Also I made the `batchRole` a public readonly property because this is then inline with the existing `role` property, and because of that thought `enableBatchBuilds()` didn't really need to return anything.

Thoughts?

Finally I updated the action integration test and added a new integration test for a batch webhook trigger.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: a85453406dff60938d6d4d5366747ebc02c6be8d
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c3c21c79-3dbe-411a-888f-bee83e6b377a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 2aa8ccf270220dc89d33e25df083ec13d83b5c5b
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=76fdd5a6-9b88-4107-bc62-8fd82b90818d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 5275bc3a62080634ad30604b72b54f6cc3070843
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=25953c0e-482d-41a8-a0fe-7efaff20138a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -44,6 +44,14 @@ export interface IProject extends IResource, iam.IGrantable, ec2.IConnectable {
   readonly role?: iam.IRole;
 
+  /** The batch IAM service Role of this Project. Undefined for imported Projects or Projects that have not configured batch builds. */
+  readonly batchRole?: iam.IRole;
```
> Also I made the batchRole a public readonly property because this is then inline with the existing role property, and because of that thought enableBatchBuilds() didn't really need to return anything.

I would get rid of this, because the value of this property changes with the lifetime of the Project instance. For this reason, I'd prefer for `enableBatchBuilds()` to return a value, like I originally suggested.

tjenkinson replied:
Cool will update 👍  It felt weird because if you wanted to do something with the batch role, if batch builds were enabled, you wouldn't be able to do that if `enableBatchBuilds()` returned the thing you needed, because the act of calling it enables it.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -196,6 +207,11 @@ abstract class ProjectBase extends Resource implements IProject {
     return this._connections;
   }
 
+  /**
+   * Enable batch builds.
+   */
```
If the JSDocs are the same, just leave them out, and they will be inherited.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -196,6 +207,11 @@ abstract class ProjectBase extends Resource implements IProject {
+  /**
+   * Enable batch builds.
+   */
+  public abstract enableBatchBuilds(): void
```
This default implementation should _not_ be abstract; it should be concrete, and just return `undefined` (otherwise, you have to repeat this default implementation twice below).

tjenkinson replied:
Cool. I actually did that initially but then switched it when I saw

```ts
  /** The IAM service Role of this Project. */
  public abstract readonly role?: iam.IRole;
```
which could also be non abstract and default to `undefined`

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -592,6 +608,15 @@ export interface ProjectProps extends CommonProjectProps {
+   *
+   * @default false
+   */
+  readonly supportBatchBuildType?: boolean;
```
> I left the supportBatchBuildType option in for now because I think it's something you also might want to enable for triggering a build manually? Otherwise you'd need to construct the project and then immediately call enableBatchBuilds().

I don't like this property, because it means there will be 2 ways to do the same thing. I don't see a problem with calling `enableBatchBuilds()` on a Project instance.

I would get rid of this.

tjenkinson replied:
Fair enough. Will remove that option.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -813,6 +853,14 @@ export class Project extends ProjectBase {
       sourceVersion: sourceConfig.sourceVersion,
       vpcConfig: this.configureVpc(props),
       logsConfig: this.renderLoggingConfiguration(props.logging),
+      buildBatchConfig: Lazy.any({
+        produce: () => {
+          const config: CfnProject.ProjectBuildBatchConfigProperty | undefined = this._batchServiceRole ? {
+            serviceRole: this._batchServiceRole.roleArn,
+          } : undefined;
+          return config;
+        },
+      }),
```
I don't see a reason for the temporary variable there.... how about something like

```suggestion
      buildBatchConfig: Lazy.any({ produce: () => this._batchServiceRole
        ? {
          serviceRole: this._batchServiceRole.roleArn,
        }
        : undefined
      }),
```

tjenkinson replied:
This was intentional to make the type checking a bit more robust. With my approach the build will fail if `config` is not the correct type.

```ts
      buildBatchConfig: Lazy.any({
        produce: () => {
          return 'a';
        },
      }),
 ```
does not error for example

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -912,7 +980,7 @@ export class Project extends ProjectBase {
       const projectStack = Stack.of(this);
       if (!(options.artifactBucket.encryptionKey instanceof kms.Key &&
-          (keyStack.account !== projectStack.account || keyStack.region !== projectStack.region))) {
+        (keyStack.account !== projectStack.account || keyStack.region !== projectStack.region))) {
```
Let's leave this double-indent as-is.

tjenkinson replied:
my editor 'fixed' these at some point. will revert

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -1221,8 +1289,8 @@ export class Project extends ProjectBase {
     if ((sourceType === CODEPIPELINE_SOURCE_ARTIFACTS_TYPE ||
-        artifactsType === CODEPIPELINE_SOURCE_ARTIFACTS_TYPE) &&
-        (sourceType !== artifactsType)) {
+      artifactsType === CODEPIPELINE_SOURCE_ARTIFACTS_TYPE) &&
```
Let's leave this double-indent as-is.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -1377,7 +1445,7 @@ class ArmBuildImage implements IBuildImage {
     const ret = [];
     if (buildEnvironment.computeType &&
-        buildEnvironment.computeType !== ComputeType.LARGE) {
+      buildEnvironment.computeType !== ComputeType.LARGE) {
```
Let's leave this double-indent as-is.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/source.ts:
```diff
@@ -531,6 +533,11 @@ abstract class ThirdPartyGitSource extends GitSource {
     const superConfig = super.bind(_scope, _project);
+
+    if (this.webhookTriggersBatchBuild) {
+      _project.enableBatchBuilds();
```
Since now `_project` has been used, rename it to `project`.

skinny85 commented on packages/@aws-cdk/aws-codebuild/test/integ.github-webhook-batch.ts:
```diff
@@ -0,0 +1,29 @@
+import * as cdk from '@aws-cdk/core';
+import * as codebuild from '../lib';
+
+class TestStack extends cdk.Stack {
```
Thanks for adding these! But I think this one integ test in the CodeBuild module is enough. I would get rid of `integ.batch.lit.ts ` (and its expected file).

skinny85 commented on packages/@aws-cdk/aws-codebuild/test/integ.github-webhook-batch.ts:
```diff
@@ -0,0 +1,29 @@
+    const source = codebuild.Source.gitHub({
+      owner: 'aws',
+      repo: 'aws-cdk',
+      reportBuildStatus: false,
```
I _think_ the default for this is `false`, so no reason to set it.

tjenkinson replied:
```ts
  /**
   * Whether to send notifications on your build's start and end.
   *
   * @default true
   */
  readonly reportBuildStatus?: boolean;
  ```
  
It's true by default, and also `false` in the `integ.github` test, so I'll leave it for now. Let me know if you still want it switching

skinny85 commented on packages/@aws-cdk/aws-codebuild/test/test.codebuild.ts:
```diff
@@ -1276,11 +1338,11 @@ export = {
 
         expect(stack).to(haveResourceLike('AWS::CodeBuild::Project', {
           'Artifacts':
-            {
```
If you're changing this, can you move the `{` to the upper line (`'Artifacts':`)?

skinny85 commented on packages/@aws-cdk/aws-codebuild/test/test.codebuild.ts:
```diff
@@ -1303,11 +1365,11 @@ export = {
 
         expect(stack).to(haveResourceLike('AWS::CodeBuild::Project', {
           'Artifacts':
-            {
```
If you're changing this, can you move the `{` to the upper line (`'Artifacts':`)?

skinny85 reviewed (changes requested):
Looks great @tjenkinson ! Some minor comments below. I see where you're coming from, but I have reasons to stay with the design I've initially proposed - I've explained them in the review comments. Hope you agree!

tjenkinson commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -156,6 +164,14 @@ export interface IProject extends IResource, iam.IGrantable, ec2.IConnectable {
+/**
+ * The return value for the {@link Project} `enableBatchBuilds()` method.
+ */
+export interface IEnableBatchBuildsReturn {
```
@skinny85 I named this this instead of `BatchBuildConfig` because I didn't want it to get confused with the actual `buildBatchConfig` in L1

skinny85 replied:
Meh. The type on the L1 level is called `ProjectBuildBatchConfigProperty`, so I'm not too worried about this one being called `BatchBuildConfig`.

The name is not a coincidence - it's actually a convention we use for return types (see examples [[1]](https://github.com/aws/aws-cdk/blob/855ce59039a577d142d68720e86d81610edffc64/packages/%40aws-cdk/aws-codebuild/lib/source.ts#L21-L32), [[2]](https://github.com/aws/aws-cdk/blob/855ce59039a577d142d68720e86d81610edffc64/packages/%40aws-cdk/aws-codebuild/lib/artifacts.ts#L12-L17)).

A name ending in `Return` seems a little weird given that, so if you could change it, I would appreciate it 🙂.

tjenkinson replied:
No problem will update then. Thanks for the examples

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c445ca70ffa78b3c18a7d46c1ab0da268eb673e2
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5409e382-1770-4c22-ae14-ad1cb8f3c526%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 26fde2e165190d2e4b1ca8989ce5d193482b1d90
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cdf35496-30a7-4a6f-b2de-8152f99da025%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 769472b9d0038ad68425f9eef95f055d1c1ad3f8
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=40c71e65-92ad-4fed-88ba-02c398656821%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c5d7ce2e10cde6c2b325841a2c57490e4bbb04f2
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=54fe0504-1aaa-48c0-8107-5c3104a19a39%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -853,6 +882,24 @@ export class Project extends ProjectBase {
+  public enableBatchBuilds(): IEnableBatchBuildsReturn | undefined {
+    if (!this._batchServiceRole) {
+      this._batchServiceRole = new iam.Role(this, 'BatchServiceRole', {
+        roleName: PhysicalName.GENERATE_IF_NEEDED,
```
Curious, why this here? Do you expect this Role to be used in a cross-env fashion?

tjenkinson replied:
Don't think so. I just copied it from somewhere else in this file 😆

Will remove

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -853,6 +882,24 @@ export class Project extends ProjectBase {
+        assumedBy: new iam.ServicePrincipal('codebuild.amazonaws.com'),
+      });
+      this._batchServiceRole.addToPrincipalPolicy(new iam.PolicyStatement({
+        resources: [Lazy.string({
```
Why use a Lazy here? Why not just `resources: [this.projectArn],`?

tjenkinson replied:
Because when bind() is called on the sources we haven't calculated the name yet meaning it was undefined. It was a bug I didn't spot earlier.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -853,6 +882,24 @@ export class Project extends ProjectBase {
+        resources: [Lazy.string({
+          produce: () => this.projectArn,
+        })],
+        actions: ['codebuild:StartBuild', 'codebuild:StopBuild', 'codebuild:RetryBuild'],
```
Can you format these on separate lines, so they're clearly visible?

```suggestion
        actions: [
          'codebuild:StartBuild',
          'codebuild:StopBuild',
          'codebuild:RetryBuild',
        ],
```

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -853,6 +882,24 @@ export class Project extends ProjectBase {
+          produce: () => this.projectArn,
+        })],
+        actions: ['codebuild:StartBuild', 'codebuild:StopBuild', 'codebuild:RetryBuild'],
+      }));
```
Don't we need some batch permissions here? Like this? https://github.com/aws/aws-cdk/pull/12181/files?file-filters%5B%5D=.ts#diff-7902b8e1bfd8571028c1c97116a58c4f6e090f92d7df3490e16e12defbfa2fab

tjenkinson replied:
I don't think so. The docs don't mention it and it worked when I tested this in a different account.

Think this is so that internally non batch builds can be triggered as part of a batch. Don't think a batch can trigger another batch?

skinny85 reviewed (changes requested):
Looks great @tjenkinson ! One more rename to go, and some clarifying questions. But this will be shipped very soon!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 54812548c09511a94f28b8a2c83959d1ade58d83
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=01a6ee1b-326e-4555-958f-6ce55916969f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: aeda32a381a35d6f562421732c0ffd63032ed57a
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ea4944e6-c308-463d-8491-6613b0a8887a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 0d7d183343e27aa09e5babbadf2877887dde4928
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5e7ea387-bfca-47c7-b328-3137524405ae%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 1a1244b4dfb48c44b54eec062578596b286d834c
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0cbeb61e-5876-4fc0-81ea-0fa646222e03%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 reviewed (approved):
Thanks so much for the contribution and working on the PR iterations @tjenkinson!

tjenkinson commented:
No problem thanks for the help! Hopefully this is the last PR we need to get batch builds working 🤞

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ecf030d7296a9519112677d019b47f0d3cdbfff8
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d99c878f-0c86-47df-8b22-8d08fded619f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 339e0eab47da33feefca1a14fcd45d4a42047479
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a0bf5425-2cba-4f26-8fe6-d1c7aee0f4ee%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
